### PR TITLE
2/3 feat(elastic_agent) add GPU statistic report to resource monitor.

### PR DESCRIPTION
**Pull request by Antlera**
### What changes were proposed in this pull request?

Add GPU statistic report to resource monitor by modifying report_resource.

### Why are the changes needed?

We need to report GPU statistics by grpc request to pass training statistics to master.

### Does this PR introduce any user-facing change?

NO.

### How was this patch tested?

Tests are performed using the included unit test code.

### CODE REVIEW

File: dlrover/python/elastic_agent/master_client.py [hunk 1]
```diff
@@ -194,12 +194,19 @@ def report_shard_checkpoint(self, shard_checkpoint):
         return self._stub.report_shard_checkpoint(request)
 
     @retry_grpc_request
-    def report_used_resource(self, memory, cpu):
+    def report_used_resource(self, memory, cpu, gpu_stats):
         request = elastic_training_pb2.ReportUsedResourceRequest()
         request.memory = memory
         request.cpu = cpu
+        for gpu in gpu_stats:
+            gpu_stats_message = request.gpu_stats[gpu["index"]]
```
Comment (workingloong) [inline dlrover/python/elastic_agent/master_client.py:202]: gpu_stats should be `list[GPUMetric]` and `gpu["index"]` should be `gpu.index`?

File: dlrover/python/elastic_agent/monitor/resource.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def get_gpu_stats(gpus=[]):
     if not gpus:
         device_count = pynvml.nvmlDeviceGetCount()
         gpus = list(range(device_count))
-    gpu_stats_list = []
+    gpu_stats_list: list[GPUMetric] = []
```
Comment (workingloong) [inline dlrover/python/elastic_agent/monitor/resource.py:63]: `gpu_stats_list` -> `gpu_stats`.  Generally, the variable name does not contain the data type.

File: dlrover/python/elastic_agent/monitor/resource.py [hunk 3]
```diff
@@ -164,8 +165,10 @@ def report_resource(self):
         try:
             used_mem = get_used_memory()
             cpu_percent = get_process_cpu_percent()
+            print(f"_gpu_enabled: {self._gpu_enabled}")
```
Comment (workingloong) [inline dlrover/python/elastic_agent/monitor/resource.py]: Replace `print` with `logger.info`.

File: dlrover/python/elastic_agent/monitor/resource.py [hunk 4]
```diff
@@ -164,8 +165,10 @@ def report_resource(self):
...
             if self._gpu_enabled:
                 self._gpu_stats = get_gpu_stats()
+            print(f"gpus: {self._gpu_stats}")
```
Comment (workingloong) [inline dlrover/python/elastic_agent/monitor/resource.py]: Replace `print` with `logger.info`.

File: dlrover/python/elastic_agent/monitor/resource.py (hunk 2 above)
Comment (Antlera) [reply dlrover/python/elastic_agent/monitor/resource.py]: Ok, thanks for the correction. I'll keep an eye on this naming issue afterward as well.

File: dlrover/python/elastic_agent/master_client.py (hunk 1 above)
Comment (Antlera) [reply dlrover/python/elastic_agent/master_client.py]: Yes. Rewrited this with GPUMetric.

File: dlrover/python/elastic_agent/master_client.py [hunk 5]
```diff
@@ -194,12 +194,19 @@ def report_shard_checkpoint(self, shard_checkpoint):
         return self._stub.report_shard_checkpoint(request)
 
     @retry_grpc_request
-    def report_used_resource(self, memory, cpu):
+    def report_used_resource(self, memory, cpu, gpu_stats):
```
Comment (workingloong) [inline dlrover/python/elastic_agent/master_client.py:197]: Fix the test case of `report_used_resource` to cover codes of gpu_stats.

**Pull request merged**